### ISSUE TITLE
[ENH] Proposal: Enforce automagic column name conversion to string type in `clean_names`

### ISSUE THREAD

**Issue by shantanuo** [ericmjl/pyjanitor#612]
UPDATE FROM MAINTAINERS: ANYBODY WHO IS INTERESTED IN THIS ISSUE, PLEASE SEE [THIS COMMENT](https://github.com/ericmjl/pyjanitor/issues/612#issuecomment-557250963) FOR PROPOSED CODE.

------

# Brief Description

`clean_names` method does not work when an integer is used as column name 

# Minimally Reproducible Code

```
rankings = {
    "countries_to_play_cricket": [
        "India",
        "South Africa",
        "England",
        "New Zealand",
        "Australia",
    ],
    0: ["England", "India", "New Zealand", "South Africa", "Pakistan"],
    "t20": ["Pakistan", "India", "Australia", "England", "New Zealand"],
}

rankings_pd = pd.DataFrame(rankings)

df = rankings_pd.clean_names()
```

# Error Messages

```
AttributeError: 'int' object has no attribute 'lower'
```

If I change the column name integer 0 to string "0" then there will be no error. But in that case it defeats the purpose of using a method :)

**Comment by rahosbach** [comment]
This is similar to #564 , which seems to have gone a bit stale at this point.

**Comment by shantanuo** [comment]
Yes. This is exact duplicate of the bug #564 
But I did not understand what "stale" mean. Is it that difficult to convert the integers to strings in python?
_____

>> I feel like because non-string column names are supported by pandas, pyjanitor should support that capability.

I do not agree with this approach. There are a lot of things pandas is not doing correctly. For e.g. issue #70 deals with invalid parameters in methods like read_excel. Pandas simply ignores the extra parameters without throwing an exception. It leads to lot of confusion. We should fix this, that and more!

**Comment by rahosbach** [comment]
@shantanuo I'm copying in your email reply here so that it's on the thread for others to review (in italics):

_Yes. This is exact duplicate of the bug #564 
But I did not understand what "stale" mean. Is it that difficult to convert the integers to strings in python?<br>
"I feel like because non-string column names are supported by pandas, pyjanitor should support that capability."<br>
I do not agree with this approach. There are a lot of things pandas is not doing correctly. For e.g. issue #70 deals with invalid parameters in methods like read_excel. Pandas simply ignores the extra parameters without throwing an exception. It leads to lot of confusion. We should fix this, that and more!_

In terms of "stale," I was simply indicating that issue #564 hasn't had a comment response in over a month.  So, it's not clear right now who (if anyone) is taking charge on it (though it does appear from his comment that @dendrondal may have started on some stuff).

@ericmjl I suggest we drop this issue and merge it in with #564.  We can continue the conversation there.  But, I'm not sure the best way to do that in GitHub (and I'm guessing that yourself or another collaborator would be the only ones with the power to do it).

**Comment by dendrondal** [comment]
#610 should have fixed this issue. I'll see later this evening if I can find out what's causing this issue. Either way, this should definitely be merged with #564 . 

@shandou , I agree with your sentiment on Pandas' approach to typing. I think it should enforce strictly string-based column names, automatically doing the conversion when the dataframe is instantiated. However, I believe the sentiment with PyJanitor is that it is meant to be a direct drop-in replacement for Pandas, meaning the same typing should be used, allowing for easy transfer of code between the two.

I briefly mentioned automatic conversion of column names to strings in #564 . I believe the quickest way to do this would be to create a wrapper around pd.DataFrame which does this, which would also allow customization of dataframe attributes down the road, if needed (i.e. for compatibility with PySpark, etc). @ericmjl , I'd love to hear your thoughts on this, as well as any other features you think would be a good inclusion to a janitor.DataFrame class.

**Comment by ericmjl** [comment]
@dendrondal and @rahosbach, thank you for chiming in in my absence. Apologies, work has taken over many hours of my day.

First off, to address the main issue: yes, we recently merged #610, though the implementation has not changed, hence the error message still shows up. To solve this, I am comfortable with accepting a PR that enforces string conversion inside of `clean_names` for every column name before name cleaning, as long as the behaviour is documented in the docstring and justifies the design choice. After all, some people will have integer column names, and to have them suddenly (and more importantly, _silently_) changed to strings is going to surprise people. The fewer the surprises the better. @shantanuo, this would be a one-line code change + docstring change + test addition, and I would love for this to be another open source contribution from you.

Now, onto some bigger issues.

The biggest one is the relationship between `pyjanitor` and `pandas`. I started it initially as a port of the `janitor` package, but as people used it and the user base grew, maybe some confusions w.r.t. the API policy. I've had it in my head for a long time, but never wrote it down formally, so this is something that should be on me to do, but I'll state the summary below:

- pyjanitor only extends or aliases the pandas API (and other dataframe APIs), but will never fix or replace them.
- Undesirable pandas behaviour should be reported upstream in the `pandas` repository.
- If at some point the `pandas` devs decide to take something from `pyjanitor` and internalize it as part of the official `pandas` API, then we will deprecate it from `pyjanitor`, while acknowledge the original contributors' contribution as part of the official deprecation record.

(If others have ideas and/or would like to officially make an API Policy section/file in our docs by first copy/pasting the above bullet points, happy to review the PR!)

Secondly, on why there's no `janitor.DataFrame` class. Actually, it did exist before, but I scrapped it once pandas-flavor came along. Subclassing the DataFrame is quite tricky to maintain, and resulted in code duplication, while pandas-flavor allowed us to avoid duplication of function names. The philosophical/design choice here was that the fewer objects with state to maintain, the better. As I've learned, functions > objects for most applications that I've encountered.

Also, big thanks to @rahosbach for addressing @shantanuo's email. I do prefer that the discussion stay on the issue tracker, so that no contributor feels overburdened to do software support alone, so what you did, @rahosbach, was the right thing in my view! Centralizing the communication channels here where the public record is visible also helps others learn about the development of the project.

Now, for some housekeeping business.

#564 - I think @dendrondal has done a tremendous job with the PR #610. So I'm going to close that one. Yes, the `clean_names` bug is still present :smile:, but hear me out.

I'm going to change the name of this issue, to propose that `clean_names` does automagic string conversion. The line change is simple, so anyone who wants to take this can do so (I haven't had the bandwidth to do a full PR, but am happy to give away the contribution to others):

```python
# if we want mapping back to originals:
new_cols = []
mapping = {}
for c in df.columns:
    new_cols.append(str(c))
    mapping[str(c)] = c
# Then use the mapping to convert back to original column name, maybe?
```

And then we'd probably want some docstrings explaining why there's conversion to strings + no conversion back to original types... unless someone wants to take that too! :smile:

I'm going to end my posting here, and get back to work :).

**Comment by ericmjl** [comment]
To all on this thread, I have just submitted PR #618 which I think will close this issue. Review requested.
